src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs

```diff
@@ -126,6 +110,40 @@ public bool HasValue
         /// <returns>The <see cref="ReadOnlySpan{T}"/> from this <see cref="StringSegment"/>.</returns>
         public ReadOnlySpan<char> AsSpan() => Buffer.AsSpan(Offset, Length);
 
+        /// <summary>
+        /// Gets a <see cref="ReadOnlySpan{T}"/> from the current <see cref="StringSegment"/> that starts
+        /// at the position specified by <paramref name="start"/>, and has the remaining length.
+        /// </summary>
+        /// <param name="start">The zero-based starting character position in this <see cref="StringSegment"/>.</param>
+        /// <returns>A <see cref="ReadOnlySpan{T}"/> with the remaining chars that begins at <paramref name="start"/> in
+        /// this <see cref="StringSegment"/>.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="start"/> is greater than or equal to <see cref="Length"/> or less than zero.
+        /// </exception>
+        public ReadOnlySpan<char> AsSpan(int start) => AsSpan(start, Length - start);
+
+        /// <summary>
+        /// Gets a <see cref="ReadOnlySpan{T}"/> from the current <see cref="StringSegment"/> that starts
+        /// at the position specified by <paramref name="start"/>, and has the specified <paramref name="length"/>.
+        /// </summary>
+        /// <param name="start">The zero-based starting character position in this <see cref="StringSegment"/>.</param>
+        /// <param name="length">The number of characters in the span.</param>
+        /// <returns>A <see cref="ReadOnlySpan{T}"/> with length <paramref name="length"/> that begins at
+        /// <paramref name="start"/> in this <see cref="StringSegment"/>.</returns>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// <paramref name="start"/> or <paramref name="length"/> is less than zero, or <paramref name="start"/> + <paramref name="length"/> is
+        /// greater than <see cref="Length"/>.
+        /// </exception>
+        public ReadOnlySpan<char> AsSpan(int start, int length)
+        {
+            if (!HasValue || (start | length) < 0 || (uint)(start + length) > (uint)Length)
```

Comment:
```suggestion
            if (!HasValue || start < 0 || (uint)(start + length) > (uint)Length)
```

`MemoryExtensions.AsSpan(this string text, int start, int length)` throws when `length` is negative, therefore this check is redundant.

Reply:
Good catch!

The difference is that `MemoryExtensions.AsSpan` throws AOR with paramName `start` only, whilst the check here distinguishes between `start` and `length` to be on par with SubSegment, and SubString.

So what to do :man_shrugging:

Reply:
OK, I missed that `paramName` was always `start` in `MemoryExtensions.AsSpan`. This seems like a bug?

`Span(T[] array, int start, int length)` just avoids specifying the `paramName`, which is maybe the best approach?

Reply:
@eerhardt do you think this is a bug in `MemoryExtensions.AsSpan`?

Reply:
@GrabYourPitchforks is the person I'd ask...

It feels like a bug to me. It definitely doesn't seem correct that `MemoryExtensions.AsSpan` and `[ReadOnly]Span.Slice` act differently here. I assume the `Span.Slice` behavior is "more correct" because you can't definitively tell which parameter is really at fault here without doing more checks which just adds code to a very hot path.

Maybe the best approach would be to open an issue describing the difference. I don't think we should try to solve it here.

Reply:
also FYI @tarekgh. Tracing back to https://github.com/dotnet/corefx/commit/a68803c848df70ae900f53e602424ba71a4760b1, it appears `Span.Slice` used `start` before that PR, but during that PR Span got split into `Fast` and `Portable`. The `Portable` path [still used `start`](https://github.com/dotnet/corefx/commit/a68803c848df70ae900f53e602424ba71a4760b1#diff-43bb305ca19d06bc64ff0599d5fc570cec8c0f04b8a462f61fc8ffd92464a816L383), but the `Fast` path dropped the parameter name.

Reply:
Filed https://github.com/dotnet/runtime/issues/53622 for this.

Reply:
The `MemoryExtensions` behavior is intentional, but I agree it's weird. We want the API to be as fast as possible, which means we don't have the luxury of figuring out which of _start_ or _length_ was incorrect. We could pass a nullptr argument name to the `ArgumentOutOfRangeException` ctor, but honestly nobody expects `ArgumentException.ParamName` to be nullptr. So always passing "start" seems like the best solution given these constraints. If it's really a problem in practice, we could always create a detailed error message, just for this one scenario.

Reply:
> We could pass a nullptr argument name to the ArgumentOutOfRangeException ctor, but honestly nobody expects ArgumentException.ParamName to be nullptr.

But it is OK for Span?

https://github.com/dotnet/runtime/blob/dbb05eba3b229df3d4c15ce08de9f84c7febb7b3/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L350-L359

Reply:
(I guess we can have this discussion in #53622 😄)